### ISSUE TITLE
Warn on overloads which become ambiguous in common Java 8 usage patterns

### ISSUE THREAD

**Issue by cushon** [google/error-prone#263]
_[Original issue](https://code.google.com/p/error-prone/issues/detail?id=263) created by **kevinb@google.com** on 2014-09-11 at 04:12 PM_

---

Consider:

```java
class A {
  static B c(D d) { ... }
  B c(A a, D d) { ... }
}
```

(I don't know whether it's even proper to call these "overloads" or not, but you get the idea.)

This is fine in Java 7 but poses a problem in Java 8. In, for example, a context which expects a `BiFunction<A,D,B>`, the method reference A::c will now be ambiguous. Whether javac gives an error or has a rule for which one to choose, it's not a good situation either way. It would be useful if compiling class A produced a warning.

I believe there is a similar case:

```java
void foo(Runnable r);
void foo(Callable<Bar> c);
```

I think there can be some calls (perhaps `foo(() -> { doSomething(); })`?) where the lambda could be successfully target-typed to match either overload, which is bad for the same reasons.

(Sorry as usual for being too rushed to get more accurate details atm.)

I would recommend producing this warning even when compiling for pre-Java8, for future-proofing.

**Comment by cushon** [comment]
This sounds like a good check. I confirmed that javac at least reports an error (instead of silently picking one):

**Ambiguous method references**

```
import java.util.function.BiFunction;

class A {
  Object c(Object d) { return null; }
  static Object c(A a, Object d) { return null; }
}

class Test {
  void f(BiFunction<A,Object,Object> f) {}
  void m() {
    f(A::c);
  }
}
```

```
error: method f in class Test cannot be applied to given types;
    f(A::c);
     ^
  required: BiFunction<A,Object,Object>
  found: A::c
  reason: argument mismatch; invalid method reference
      reference to c is ambiguous
        both method c(A,Object) in A and method c(Object) in A match
```

**Overloads accepting same functional interface types**

```
import java.util.concurrent.Callable;

abstract class Test {
  interface MyRunnable {
    void run();
  }

  abstract void foo(Runnable r);
  abstract void foo(MyRunnable r);

  void m() {
    foo(() -> { });
  }
}
```

```
error: reference to foo is ambiguous
    foo(() -> { });
    ^
  both method foo(Runnable) in Test and method foo(MyRunnable) in Test match
```

**Comment by kevinb9n** [comment]
FYI, I just had this idea again, and pitched it to Louis again, and he liked it again.

**Comment by mirabilos** [comment]
Also this:

    int skip(final Function<Integer, Boolean> fn);
    int skip(final BiFunction<Integer, Integer, Boolean> fn);

https://stackoverflow.com/a/36257750/2171120 explains how they can become conflicting/ambiguous, and javac 8 refuses to compile them but while later javac versions succeed these constructions can make the API ambiguous (especially as IntelliJ insists on replacing lambdas with method references)…
